mojoshader_metal.c

```diff
@@ -0,0 +1,996 @@
+    ubo->internalOffset = 0;
+    for (int i = 0; i < ubo->numInternalBuffers; i++)
+    {
+        ubo->internalBuffers[i] = NULL;
+        ubo->internalBuffers[i] = UBO_create_backing_buffer(ubo, i);
+    }
+
+    /* Add the UBO to the global list so it can be updated. */
+    LLNODE *node = LL_append_node(&ubos, m, d);
+    node->data = ubo;
+
+    return ubo;
+} // create_ubo
+
+static void dealloc_ubo(MOJOSHADER_mtlShader *shader,
+                        MOJOSHADER_free f,
+                        void* d)
+{
+    if (shader->ubo == NULL)
+        return;
+
+    LL_remove_node(&ubos, shader->ubo, f, d);
+    for (int i = 0; i < shader->ubo->numInternalBuffers; i++)
+    {
+        objc_msgSend(shader->ubo->internalBuffers[i], selRelease);
+        shader->ubo->internalBuffers[i] = NULL;
+    }
+    f(shader->ubo->internalBuffers, d);
+    f(shader->ubo, d);
+} // dealloc_ubo
+
+static void *get_uniform_buffer(MOJOSHADER_mtlShader *shader)
+{
+    if (shader == NULL || shader->ubo == NULL)
+        return NULL;
+
+    return shader->ubo->internalBuffers[shader->ubo->currentFrame];
+} // get_uniform_buffer
+
+static int get_uniform_offset(MOJOSHADER_mtlShader *shader)
+{
+    if (shader == NULL || shader->ubo == NULL)
+        return 0;
+
+    return shader->ubo->internalOffset;
+} // get_uniform_offset
+
+static void update_uniform_buffer(MOJOSHADER_mtlShader *shader)
+{
+    if (shader == NULL || shader->ubo == NULL)
+        return;
+
+    float *regF; int *regI; uint8 *regB;
+    if (shader->parseData->shader_type == MOJOSHADER_TYPE_VERTEX)
+    {
+        regF = vs_reg_file_f;
+        regI = vs_reg_file_i;
+        regB = vs_reg_file_b;
+    }
+    else
+    {
+        regF = ps_reg_file_f;
+        regI = ps_reg_file_i;
+        regB = ps_reg_file_b;
+    }
+
+    UBO_predraw(shader->ubo);
+    void *buf = shader->ubo->internalBuffers[shader->ubo->currentFrame];
+    void *contents = UBO_buffer_contents(buf) + shader->ubo->internalOffset;
+
+    int offset = 0;
+    for (int i = 0; i < shader->parseData->uniform_count; i++)
+    {
+        int idx = shader->parseData->uniforms[i].index;
+        int arrayCount = shader->parseData->uniforms[i].array_count;
+        int size = arrayCount ? arrayCount : 1;
+
+        switch (shader->parseData->uniforms[i].type)
+        {
+            case MOJOSHADER_UNIFORM_FLOAT:
+                memcpy(
+                    contents + (offset * 16),
+                    &regF[4 * idx],
+                    size * 16
+                );
+                break;
+
+            case MOJOSHADER_UNIFORM_INT:
+                // !!! FIXME: Need a test case
+                memcpy(
+                    contents + (offset * 16),
+                    &regI[4 * idx],
+                    size * 16
+                );
+                break;
+
+            case MOJOSHADER_UNIFORM_BOOL:
+                // !!! FIXME: Need a test case
+                memcpy(
```

Comment:
Actually, just noticed that this array should just be a u8 array, not an ivec or anything weird like that. (Unless the UBO forced it to be an ivec4?)